### PR TITLE
Add USB MIDI support to libraries/USB

### PR DESCRIPTION
**Pull request by EleotleCram**
## Description of Change
This PR adds support for USB MIDI to libraries/USB. It wraps the underlying tiny-usb API in an Arduino-like/User-friendly API, similar to the already existing USBHID* classes.

## Tests scenarios
All the examples have been tested using an ESP32-S2 Mini. Please see the related links section for video-proof that it works.

## Related links
* MIDI Music Box Demo - https://youtu.be/JFrc-wSmcus
* MIDI Interface Demo - https://youtu.be/BXG5i55I9s0
* MIDI Controller Demo - https://youtu.be/Y9TLXs_3w1M

### CODE REVIEW

**Comment by me-no-dev** [comment]
Very nice implementation!

- You need to add `USBMIDI.cpp` [here](https://github.com/EleotleCram/arduino-esp32/blob/usb-midi-support/CMakeLists.txt#L114)
- There are also some other build errors reported by CI [here](https://github.com/espressif/arduino-esp32/actions/runs/4902674538/jobs/8845612539?pr=8166#step:5:92)

**Comment by EleotleCram** [comment]
Thanks!

And I was already suspicious on how silent the _arduino-cli_ output was, I just discovered the `--warnings all` option ;-)
Will fix this shortly.

File: libraries/USB/src/USBMIDI.cpp [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+bool USBMIDI::readPacket(midiEventPacket_t *packet) {
+    return tud_midi_packet_read((uint8_t *)packet);
+}
+
+bool USBMIDI::writePacket(midiEventPacket_t *packet) {
+    return tud_midi_packet_write((uint8_t *)packet);
+}
+
+// Default Cable Number (for simplified APIs that do not expose this)
+#define DEFAULT_CN 0
```
Comment (lucasssvaz) [inline libraries/USB/src/USBMIDI.cpp:130]: Wouldn't it be better to move this to the beginning of the file ?

File: libraries/USB/src/USBMIDI.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    ///// USB MIDI 1.0 interface /////
+    //
+    // Attempt to read a USB MIDI packet from the USB Bus
+    bool readPacket(midiEventPacket_t *packet);
+    // Attempt to write a USB MIDI packet to the USB Bus
+    bool writePacket(midiEventPacket_t *packet);
+
+    /////////////////////////////////////
+    ///// Serial MIDI 1.0 interface /////
+    //
```
Comment (lucasssvaz) [inline libraries/USB/src/USBMIDI.h:67]:
```suggestion
    /* Serial MIDI 1.0 interface */

```

File: libraries/USB/src/USBMIDI.h [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    // Pitch Bend Change [-8192,0,8191]
+    void pitchBend(int16_t pitchBendValue, uint8_t channel = 1);
+    // Pitch Bend Change [0,8192,16383]
+    void pitchBend(uint16_t pitchBendValue, uint8_t channel = 1);
+    // Pitch Bend Change [-1.0,0,1.0]
+    void pitchBend(double pitchBendValue, uint8_t channel = 1);
+
+    //////////////////////////////////
+    ///// USB MIDI 1.0 interface /////
+    //
```
Comment (lucasssvaz) [inline libraries/USB/src/USBMIDI.h:59]:
```suggestion
    /* USB MIDI 1.0 interface */

```

File: libraries/USB/src/USBMIDI.h [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+// Patch MIDI definition bug pending update/pull from upstream tinyusb repo.
+// See also:
+//            - https://github.com/hathach/tinyusb/pull/1920
+//            - https://github.com/espressif/arduino-esp32/pull/8117
+//
+#undef MIDI_CIN_NOTE_ON
+#define MIDI_CIN_NOTE_ON 9
+#undef MIDI_CIN_NOTE_OFF
+#define MIDI_CIN_NOTE_OFF 8
```
Comment (lucasssvaz) [inline libraries/USB/src/USBMIDI.h:17]: This is already merged upstream, could this be deleted ?

File: libraries/USB/src/USBMIDI.h [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+class USBMIDI {
+public:
+    USBMIDI(void);
+    void begin(void);
+    void end(void);
+
+    //////////////////////////
+    ///// User-level API /////
+    //
```
Comment (lucasssvaz) [inline libraries/USB/src/USBMIDI.h:37]:
```suggestion
    /* User-level API */

```

**Review by lucasssvaz — APPROVED** [review]
Working as expected

File: libraries/USB/src/USBMIDI.cpp [hunk 6]
```diff
@@ -0,0 +1,140 @@
+#include "USBMIDI.h"
+
```
Comment (P-R-O-C-H-Y) [inline libraries/USB/src/USBMIDI.cpp:2]: Please add check for USB peripheral support as other USB files has.

```suggestion
#if SOC_USB_OTG_SUPPORTED

```

File: libraries/USB/src/USBMIDI.cpp [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+size_t USBMIDI::write(uint8_t c) {
+    // MIDI_CIN_1BYTE_DATA => Verbatim MIDI byte-stream copy
+    // (See also Table 4-1 of USB MIDI spec 1.0)
+    midiEventPacket_t packet = {DEFAULT_CN | MIDI_CIN_1BYTE_DATA, c, 0, 0};
+
+    return tud_midi_packet_write((uint8_t *)&packet);
+}
+
+#endif // CONFIG_TINYUSB_MIDI_ENABLED
```
Comment (P-R-O-C-H-Y) [inline libraries/USB/src/USBMIDI.cpp:140]: Close the #if from above comment
```suggestion
#endif /* CONFIG_TINYUSB_MIDI_ENABLED */
#endif /* SOC_USB_OTG_SUPPORTED */
```

File: libraries/USB/src/USBMIDI.h [hunk 8]
```diff
@@ -0,0 +1,72 @@
+
```
Comment (P-R-O-C-H-Y) [inline libraries/USB/src/USBMIDI.h:1]: Include the `social_caps.h` and add a check also to header file.

```suggestion
#include "soc/soc_caps.h"
#if SOC_USB_OTG_SUPPORTED
```

File: libraries/USB/src/USBMIDI.h [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    bool writePacket(midiEventPacket_t *packet);
+
+    /////////////////////////////////////
+    ///// Serial MIDI 1.0 interface /////
+    //
+    // Write a Serial MIDI byte (status or data) to the USB Bus
+    size_t write(uint8_t c);
+};
+
+#endif // CONFIG_TINYUSB_MIDI_ENABLED
```
Comment (P-R-O-C-H-Y) [inline libraries/USB/src/USBMIDI.h:72]: Close the #if from above comment.
```suggestion
#endif /* CONFIG_TINYUSB_MIDI_ENABLED */
#endif /* SOC_USB_OTG_SUPPORTED */
```

**Review by P-R-O-C-H-Y — CHANGES_REQUESTED** [review]
@EleotleCram PTAL on my comments.
Also please add skip files to examples for esp32c6 and esp32h2. 
Thanks! Good job :)

**Comment by VojtechBartoska** [comment]
Hello @EleotleCram,

thanks for you PR.

Unfortunately it seems that CLA form did not show up for your first contribution. Can you please check https://cla-assistant.io/ if you have the CLA signed for this repository.

Thanks a lot

**Comment by EleotleCram** [comment]
@VojtechBartoska AFAIK I already signed the CLA for this repository; when I log in to cla-assistant and I view my signed CLAs it lists this repo.

**Pull request merged**